### ISSUE TITLE
Arg Types (Entities) API

### ISSUE THREAD

**Issue by JDGrimes** [WordPoints/hooks-api#121]
_For posterity: this ticket is created from discussion which originally took place in #13 much earlier._
## Background

In #13 we are [struggling with exactly how to relate posts and terms](https://github.com/WordPoints/hooks-api/issues/13#issuecomment-118419841). That's when the idea of an arg types API came to mind.
## Basic Concept

We could split args into Entities, Attributes, and Relationships.

An Entity would be a post, comment, or user.
An Attribute would be the content of a post or the name of a user.
A Relationship would be between a comment and a post, or a post author and a post.

**Comment by JDGrimes** [comment]
_Moved from https://github.com/WordPoints/hooks-api/issues/13#issuecomment-124887384._

## Relationship Bi-directionality

Should relationships work both directions?

We can get all authors of a post, but should we also be able to get all posts of
an author?

We would still have one-to-one and one-to-many relationships. A `post_author`
relationship would still be one-to-many (one user to many posts). We'd be able to
get the user based on a post, and all of the posts based on a user.

Specifically, we need to be able to describe the relationship for retroactive
awards in that schema language. We also need to be able to retrieve the author of
a post when a hook is fired. Whether we might also need to retrieve the posts of a
user remains to be seen. It is unlikely that we will need that, because users are
our targets at the moment. However, for other relationships (comment to post, for
example), we certainly might.

So I think we should make it bi-directional.

If we do that, we can't very well say "this is the primary entity, and this is the secondary one". So we can't have a method like `get_primary_from_secondary()`. We'd need to do something like `get_related( $type, $value )`, where type would be either 'post' or 'user', and the value would be the post ID or user ID.

Also, as I said above, we can't just say "get it from this field", because sometimes there will be multiple tables involved.

**Comment by JDGrimes** [comment]
_Moved from https://github.com/WordPoints/hooks-api/issues/13#issuecomment-125249274._

## One-to-many Relationships

I'm having trouble figuring out how to handle one-to-many relationships. Specifically, I need to think through how they will be handled in the arg hierarchy.

Take for example a Post Publish hook that has a constraint on the post's terms. The constraint could require the post to have at least x terms. Or, the constraint could require that the post have a specific term.

Normally, constraints are for entity attributes. In this case however, the constraint would be on the relationship.

In the case of a one-to-many relationship like this, we can't really have constraints directly on the attributes of the other entities in the relationship. We need to treat it as if the constraint were on the relationship itself, but it is possible that it could include certain conditions based on the attributes of the related entities. In any case, we will need to treat this as an array of entities to which the constraint(s) are applied collectively. In other words, we cannot descend into the related entities and their attributes and relationships except in the context of the constraint itself.

**Comment by JDGrimes** [comment]
_Moved from https://github.com/WordPoints/hooks-api/issues/13#issuecomment-125251112._

## Entity Arrays

We must now ask the question of how we should represent arrays of entities. I am thinking that we may want to represent them as an object.

Note that entity arrays will probably have application beyond one-to-many relationships. Imagine a scenario where a hook is fired with not just one post, but an array of posts. Usually, in such a case, the posts might have some sort of relationship. But think of a case where they do not. This is an example of a place where we would want to use an entity array as the base arg of the hook.

**Comment by JDGrimes** [comment]
_Moved from https://github.com/WordPoints/hooks-api/issues/13#issuecomment-125604581._

> In the case of a one-to-many relationship like this, we can't really have constraints directly on the attributes of the other entities in the relationship. We need to treat it as if the constraint were on the relationship itself, but it is possible that it could include certain conditions based on the attributes of the related entities. In any case, we will need to treat this as an array of entities to which the constraint(s) are applied collectively. In other words, we cannot descend into the related entities and their attributes and relationships except in the context of the constraint itself.

Actually, this has nothing to do with relationships. It is about entity arrays.
We can descend into relationships, but for a one-to-many relationship we would be
descending into an entity array. It is the entity array that we cannot descend
into.

**Comment by JDGrimes** [comment]
_Moved from https://github.com/WordPoints/hooks-api/issues/13#issuecomment-125618074._

## Entity Arrays and Constraints

Constraints we might want on entity arrays would include:
- The number of entities.
- Whether any of the entities match a set of sub-constraints.

Let's consider a practical example. We might for example want a constraint that specifies that a post have at least three terms in a specific taxonomy, e.g., 3 tags.

In that case, what we'd have would sort of be a combination of the above two example constraints types. But it wouldn't just be that there would be two constraints, like this:
1. Entity array must contain at least 3 items.
2. Entity array items must have the taxonomy "post_tag".

The reason is that there are allowed to be terms of other types as well. The post just has to match the constraint in this case, it doesn't have to be exactly equal to it.

What we have here isn't just these two constraints being used in combination. It is a single constraint which is a combination of the two.

This sort of composite constraint is composed of two different basic types of conditions:
1. Conditions on the number of items that must match.
2. Conditions on the attributes of those items.

These need to be applied in reverse order. First, we need to filter all of the items through the "sub-constraints" on their attributes and relationships. Then we need to check the filtered set against the numerical condition(s).

**Comment by JDGrimes** [comment]
_Moved from https://github.com/WordPoints/hooks-api/issues/13#issuecomment-125967593._

## Relationship Bi-directionality Revisited

> So I think we should make it bi-directional.
> 
> If we do that, we can't very well say "this is the primary entity, and this is the secondary one". So we can't have a method like get_primary_from_secondary(). We'd need to do something like get_related( $type, $value ), where type would be either 'post' or 'user', and the value would be the post ID or user ID.

While this makes some sense, it presents problems in one-to-many relationships, because the arg type becomes ambiguous in the API. The API doesn't know which side of the relationship is an array of entities. This is problematic when, for example, we are validating constraints. A constraint that only works on entity arrays won't validate, because the API will just pass it an entity of the correct type instead of an entity array for that type.

My proposed solution is to make constraints one directional. This has other benefits as well, such as giving more granularity to the API.

**Comment by JDGrimes** [comment]
_Moved from https://github.com/WordPoints/hooks-api/issues/13#issuecomment-125996320._

## Multi-arg type Relationships

Up to now I have been making the assumption that there will never be any more than 2 different types of args involved in a single relationship.

An arg of type A could thus only be related to:
- Another arg of type A.
- Another arg of type B.
- An array of other args of type A.
- An array of other args of type B.

I have also been assuming that entity arrays will always contain args of just one type.

I think we should consider each of these assumptions in more depth.

Let's start with multi-type entity arrays. The current design for entity arrays requires that the contain only a single type of argument. ~~For example, a constraint cannot be applied to a multi-type entity array, because the conditions~~ Actually, both the count constraints and condition constraints could be applied to a multi-type entity array. The conditions for one type in the array would be applied only to the entities of that type. We would also want to be able to have count conditions apply to each of the types (or really, we should let them apply to each of the sub-constraints).

The question is, do we really need multi-type entity arrays? The answer is: maybe. We might want to use them for the same reason that we'd want to use entity arrays in the first place: when some args come in a group but they aren't otherwise related. We could think of them sort of like run-time–only relationships. This is already true when we are using them in one-to-many relationships to represent the "many". Those args aren't necessarily related in any other way, but they happen to be currently related through their common relationship with the arg.

We also might want to use multi-type entity arrays in the case of an action that wants to fire with a group of args instead of just one. However, that seems like a real edge-case right now, so we might not want to pursue it.

Now let's consider whether we'd ever need multi-type arrays for multi-type relationships. By this I don't mean a relationship between a post and the author _and_ the comments. I mean a relationship between a post and a bunch of other things of various types that are all stored together in a single table. That really wouldn't work though, because there would be no way to tell which IDs belonged to which types of things. This is a problem with multi-type entity arrays in general. So let's not support them.

**Comment by JDGrimes** [comment]
_Moved from https://github.com/WordPoints/hooks-api/issues/13#issuecomment-140743956._

## Children and Parents

The args API supports different types of args: entities, attributes, and relationships. In trying to pin down exactly how the API should work, we need to think through how these can be classified based on whether they are children of other args, parents of other args, or both.

The importance of how different args relate to each other is that they are usually handled within a hierarchy. Knowing how to get the child of the arg is important when we are descending down the hierarchy. Currently, we do not ever ascend up the hierarchy, starting from lower down. Indeed this would be impossible to do when starting from an attribute, for example, because the attributes are only associated with their parents in the registry, and so the parent must be known to even get the attribute object. We have also decided that relationships will be unidirectional, and can therefore safely ignore the possibility of ascending up hierarchies.

It is manifest that we need to be able to get a child from the parent. There are two different workflows here, though: one, in which we need the child arg object itself, and a second, in which we are only interested in getting the child value. The first thing that we must decide is whether or not to split these two workflows to use different logic.

To that end, we should define exactly what those flows would involve.

The object flow will necessarily require knowing the parent and child args' slugs, and then retrieving the object from the correct registry. Which registry the child is in will be determined by what type of child it is, and what type the parent is. The current nature of slugs is such that they do not, in themselves, reveal this information, any more that the number 345 reveals that it is the ID of a post and not of a user; in truth it could be either, or even both. Therefore we need to somehow learn the types of the child and parent from the slugs, by looking in the various registries. However, we should note that if we have to look in the registries for several different types, this will lead to naming conflicts.

We currently always know the type of the parent, because only an entity can currently be used as the top of a hierarchy, and we keep track of the current type as we descend.

However, that still leaves the possibility of naming conflicts between children of different types. But is that really a problem? Would it ever make sense, for example, to have an attribute and relationship of the same name on the same arg? No, not really. Relationships can only have entities as children, and so there is no problem there, because entities all share a single namespace. There is a possible issue with naming conflicts when two different entities have relationships with the same name, if all relationships use a single namespace, as they currently do. I think that is a good argument for making relationships second class citizens, like attributes. The only reason to have them be first class is if we need to retrieve them directly, but that isn't currently the case, since they aren't bi-directional. (Done!)

OK, naming conflicts taken care of, we return to the child object retrieval flow. We start out knowing at least the parent slug, the parent type, and the child slug, and we go to the registries to find a child. Currently there is a separate registry for the atts and the relationships, but is that necessary? Now that they are both just children of args, why can't they all reside in a single registry? With a single registry for all entity children of all types, we no longer have to worry about adding extra logic across several registries to avoid naming conflicts. It would also make retrieving children very simple, with no extra logic required for checking in a bunch of different registries. Let's merge them together. (Done!)

In short, I see no reason to separate these flows. The parent arg should just automatically set the value of the child arg when the child is retrieved (if the parent value is set, that is). (Done!)

This work has greatly simplified the logic for handling arg hierarchies, and also makes the API even more extensible and composable. Awesome!

**Comment by JDGrimes** [comment]
_Moved from https://github.com/WordPoints/hooks-api/issues/13#issuecomment-142439749._

## Multi Arg Actions

> The question is, do we really need multi-type entity arrays? The answer is: maybe. We might want to use them for the same reason that we'd want to use entity arrays in the first place: when some args come in a group but they aren't otherwise related. We could think of them sort of like run-time–only relationships. This is already true when we are using them in one-to-many relationships to represent the "many". Those args aren't necessarily related in any other way, but they happen to be currently related through their common relationship with the arg.
> 
> We also might want to use multi-type entity arrays in the case of an action that wants to fire with a group of args instead of just one. However, that seems like a real edge-case right now, so we might not want to pursue it.
> 
> Now let's consider whether we'd ever need multi-type arrays for multi-type relationships. By this I don't mean a relationship between a post and the author and the comments. I mean a relationship between a post and a bunch of other things of various types that are all stored together in a single table. That really wouldn't work though, because there would be no way to tell which IDs belonged to which types of things. This is a problem with multi-type entity arrays in general. So let's not support them.

I'd like to revisit this idea of run-time relationships. The argument against supporting multi-type entity arrays still holds. However, in regards to multi-arg actions, I think that there may be a case to be made for them.

Multi-arg actions would be actions that relate to more than one entity. Most actions, like Post Publish, relate to just one entity (in that case, the Post). However, I'm wondering whether some actions might indeed relate to several entities. WordPress actions do, after all, accept a number of arguments, so what is to say that we don't want to relate the hook action to more than one entity if multiple ones are supplied?

An example of an action that is related to more than one entity is the Comment Approve action. The action relates to both the Comment and the Post. In this case, of course, it just happens that the Comment already has a natural relationship with the post in the database, so we only need to pass along a Comment entity to have access to the Post, the Comment Author, etc.

The question I want to explore again is whether there might be times when an action fires, and it relates to several entities that are not otherwise related.

Before when we explored this question, I was thinking only in terms of entity arrays, specifically multi-type entity arrays. And the idea of multi-type entity arrays was discarded, at least for the present. However, what I am proposing here has nothing to do with entity arrays.

First, let's get something straight: Actions actually don't pass an entity on to the event. They pass an arg hierarchy, initialized with a single entity at the top.

So when I'm talking about multi-arg actions, I'm not talking about passing an entity array instead of a single entity. I'm talking about passing a hierarchy with multiple top-level args instead of just one.

Note that this would still only allow for only one entity of each type to be passed—we're not talking about entity arrays here. Each level of the hierarchy can only have one arg of a given type.

But does that actually cripple the usefulness of such a feature?

Take an example, where we have one user editing another user. We can only include one of them in the hierarchy, not both.

To overcome that, we'd need to alias any top-level args of the same type. But that becomes fragile when we switch out actions for an event, unless the aliases are standardized across those actions. There would also be the problem of running into conflicts between aliases and other real args, unless alias slugs can be differentiated in some way.

Possibly the largest problem with this idea is how to communicate to the user in the UI how each of the other entities relates to the action. For the example of one user editing another, we can't just say we have a User and another User. Instead, we might call them the User and the Editor, for example. But in that case, we are going to need to have some sort of formal representation of that run-time relationship in the API.

This is where paths cross with our [discussion on how to handle stateful entities in the UI](https://github.com/WordPoints/hooks-api/issues/15#issuecomment-142095846). Because in fact, stateful entities are just one type of run-time relationship, so to speak.

The question then, is how should an action define the entities that it relates to and the roles that they play. Or, should that by partly defined at the event level?

That is assuming, of course, that we even need to support having actions relate to multiple entities. Do we? Of course there are the stateful entities, but is there really anything of substance beyond that?

I just skimmed through all of the over 1900 hooks in WordPress core. Most of them aren't something that we'd ever want to connect to an event. Of those that are, the following are significant:
- [`switch_theme`](https://developer.wordpress.org/reference/functions/switch_theme/) The previous theme is not provided as a hook argument, but it is stored in the `theme_switched` option. But would we ever need that?
- [`set_user_role`](https://developer.wordpress.org/reference/hooks/set_user_role/) The previous roles are passed as a hook argument. However, it is an array. There are also caveats here, since there is also an `add_role()` method with its own hook.
- [`profile_update`](https://developer.wordpress.org/reference/hooks/profile_update/) The old user data is passed as a hook argument. However, it isn't exactly an entity, just a plain object.
- [`password_reset`](https://developer.wordpress.org/reference/hooks/password_reset/) The new password is passed as a hook argument. Not exactly an entity, but we could make it one.

None of these is particularly compelling as a mainstream feature, however, I can see how all of them could prove useful. Let's not get to involved in thought about actually implementing these specifically, the point is that such things can indeed be useful.

So we should make sure that the API is flexible enough that, at the very least, we won't break all sorts of things if we try to make multi-arg hooks possible in the future. And because we are already dealing with stateful args, I think there is a compelling case to be made for going beyond that, and actually baking it into the initial API.

**Comment by JDGrimes** [comment]
_Moved from https://github.com/WordPoints/hooks-api/issues/13#issuecomment-142608967._

At present, what type of entity an Action initializes the hierarchy with is determines solely by the Action object itself. It isn't possible to even find out what type of entity the Action is associated with without retrieving an instance of the Action object.

But now I'm wondering if maybe this should change. Perhaps there should be a registry, where entities can be added to or removed from an Action. That would allow us to much more easily find the entity(ies) that an action was associated with. 

However, this brings us back around to one of the concerns which we voiced above: does this make the API too fragile?

Would the actions become less composible now that they may need to supply multiple entities to maintain parity with each other? The answer is maybe. On the one hand, this is not as much of a problem as it might otherwise be, because each of the entities could be provided by code outside of the action itself. So, for example, the code that retrieved the current user for one action would do the same for another. On the other hand, it is possible that other, non-stateful args might not be available across actions.

The real question is, is that important? Of course if you switch out actions they need to have some amount of feature parity, that is just a given. This might make it harder to do that, but it is something that would probably only be done in extreme situations anyway. I'd say it is much more edge-case than multi-arg Actions, and that is what is more important here.

Just so we're clear though, let's consider what would happen if we switched out an action that had an entity to another one that didn't. Since the API is composible, not much would happen of significance until the event was actually fired. At that point, any reactions in the database that referenced the missing entity would just fail to hit the reactor. And that is the expected behavior; nothing bad happens, just some things that happened before won't happen anymore.

Now that we've dealt with that, we need to decide whether we want to make such a feature part of the initial API, and if so, to what degree.

First, let's consider what might break were we to implement this in the future. The hierarchies, as stored in the database, already support having multiple args at the top. However, we need to:
- [x] Make sure that the extensions don't break when multiple args are at the top.
- [x] Change the arg hierarchy class to support this. Currently it can only be constructed with a single entity.
- [x] Make sure that the reactors properly handle this as well.

After that we should be ready to move forward on this bit of the API, either now or later.

**Comment by JDGrimes** [comment]
_Moved from https://github.com/WordPoints/hooks-api/issues/13#issuecomment-142652531._

### Actions and Main Entities

I think that even multi-arg actions usually have a single entity that would be considered the primary entity for the action. This is significant, especially when it comes to reversals.

When we considered action parity above, I was only thinking of actions that fired an event, not parity between the action firing an event and the one reversing it. The question is, do the firing action and the reversing action (if present) need to have parity in what entities they supply? Or do only the "main entities" (if such a thing really exists) need to match?

Actually, the question is irrelevant for most multi-arg actions, because the args are stateful in some way, and so the action isn't reversible. There is no opposite of visiting the site. There is no opposite of a password reset (you can reset the password again, but you can't un-reset it). So, for the most part this is moot.

Think of a case like this though: a user edits a post, but isn't the author. That action can be reversed if that post revision is reverted. There would need to be some special logic to catch that, since there is no simple hook we can hook into. But you get the idea. Now, the reversal is a cross-section of the user and post entities. However, we can only reverse because we know the user is associated with the revision, and in that case, we don't need to have multiple args involved in a run-time relationship at all, because the two are in fact related in the database. In fact, we'd probably set up our firing action to pass the revision entity in the first place.

So, if the action is reversible at all, the args will have to be related in the database. And in that case, both the firing action and reversing action will be an entity that embodies in some way that relationship. Meaning, of course, that there will therefore be a natural parity across the action and reverse action.

I just realized that we've never really discussed whether parity across the action(s) and reverse action(s) for an event is/should be a requirement of the API. However a survey of the API shows that all of our current events use actions with arg parity. The points reactor is also designed based on the assumption that there will be a single primary arg for each action. However, it would be possible for that to change if needed. It also assumes that the entity type will match the type of the entity which it was originally fired with. This is necessary to be able to reverse only transactions related to the that entity. 

The question is, what about when the firing action offers several entities, but the reversal action offers just one? Would the expected behavior be to reverse all transactions related to that entity, whatever other entities they were related to? I think the answer is yes.

That is the sense in which I am talking about primary entities. The entities which are so integral to the event that it can be reversed solely based on them. The other entities may be used as the target or in conditions, etc.; but they don't represent the event itself.

In that case, we don't have to have just one primary entity for an event, we could have several. But the reversal actions could still just offer only one or the other of those primary entities.

Again though, this is mostly moot, since many multi-arg actions are stateful, and stateful events by their very nature can't be reversed.
